From 3489540cdb1a24b19e32a5025d0baeaacd44e952 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Fri, 21 Nov 2025 18:56:37 +0100
Subject: [PATCH] usb: dwc2: hcd: ensure device disconnects, when swapping role

When swapping from host to device role, there may be some
race that avoids to see a disconnect. The bus may get resumed, to exit
power saving modes in drd.c. So the bus is no longer suspended; and the
device is still visible after the DR swap, with lsusb.
So take care to resume the root hub, to see a logical disconnect.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: Icb91b88134a84416c6be654f4781fb3932de4cb1
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/512981
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/usb/dwc2/hcd.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/drivers/usb/dwc2/hcd.c
+++ b/drivers/usb/dwc2/hcd.c
@@ -1816,7 +1816,8 @@ void dwc2_hcd_disconnect(struct dwc2_hso
  */
 static void dwc2_hcd_rem_wakeup(struct dwc2_hsotg *hsotg)
 {
-	if (hsotg->bus_suspended) {
+	/* when swapping role to device, need to resume rhub to see disconnect */
+	if (hsotg->bus_suspended || dwc2_is_device_mode(hsotg)) {
 		hsotg->flags.b.port_suspend_change = 1;
 		usb_hcd_resume_root_hub(hsotg->priv);
 	}
